crates/provider/src/heart.rs

```diff
@@ -458,37 +504,112 @@ impl<S> Heartbeat<S> {
     /// Spawn the heartbeat task, returning a [`HeartbeatHandle`].
-    pub(crate) fn spawn(self) -> HeartbeatHandle {
+    pub(crate) fn spawn<T: Transport + Clone>(self, client: WeakClient<T>) -> HeartbeatHandle<N> {
         let (latest, latest_rx) = watch::channel(None::<Block>);
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
-impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
-    async fn into_future(
+type FetchReceiptFuture<R> = Pin<Box<dyn Future<Output = Option<R>> + Send>>;
+
+struct ReceiptTasks<T, N: Network> {
+    client: WeakClient<T>,
+
+    /// Receiver of shutdown signal, used on unexpected channel close.
+    shutdown_tx: mpsc::Sender<()>,
+
+    /// Futures which perform a single request to fetch a receipt.
+    fetch_receipt_tasks: FuturesUnordered<FetchReceiptFuture<N::ReceiptResponse>>,
+
+    /// Futures which poll for receipts of confirmed transactions and notify [TxWatcher]s.
+    poll_receipt_tasks: FuturesUnordered<Pin<Box<dyn Future<Output = ()> + Send>>>,
+}
+
+impl<T: Transport + Clone, N: Network> ReceiptTasks<T, N> {
+    fn new(client: WeakClient<T>, shutdown_tx: mpsc::Sender<()>) -> Self {
+        Self {
+            client,
+            shutdown_tx,
+            fetch_receipt_tasks: Default::default(),
+            poll_receipt_tasks: Default::default(),
+        }
+    }
+
+    fn fetch_receipt(&mut self, tx_hash: B256) {
+        // We don't actually poll, but use poller to retry on failures.
+        let mut poller = PollerBuilder::<_, _, Option<N::ReceiptResponse>>::new(
+            self.client.clone(),
+            "eth_getTransactionReceipt",
+            (tx_hash,),
+        )
+        .spawn()
+        .into_stream();
+
+        let shutdown = self.shutdown_tx.clone();
+
+        self.fetch_receipt_tasks.push(Box::pin(async move {
+            if let Some(receipt) = poller.next().await {
+                receipt
+            } else {
+                let _ = shutdown.send(()).await;
+                None
+            }
+        }));
+    }
+
+    fn poll_receipt(&mut self, watcher: TxWatcher<N>) {
+        let mut poller = PollerBuilder::<_, _, Option<N::ReceiptResponse>>::new(
+            self.client.clone(),
+            "eth_getTransactionReceipt",
+            (watcher.config.tx_hash,),
+        )
+        .spawn()
+        .into_stream();
+
+        let shutdown = self.shutdown_tx.clone();
+
+        self.poll_receipt_tasks.push(Box::pin(async move {
+            while let Some(receipt) = poller.next().await {
+                if let Some(receipt) = receipt {
+                    watcher.notify(receipt);
+                    return;
+                }
+            }
+
+            // If the stream closed before we got a receipt, it's not OK.
+            let _ = shutdown.send(()).await;
```

Comment:
this shuts down the heartbeat?

Reply:
yeah, we send shutdown only when poller unexpectedly stops which usually means that there was either irrecoverable error or client was dropped

not sure how we can handle this differently, we can probably send `Result` through `TxWatcher` if something goes wrong